internal/xds/rbac/rbac_engine_test.go

```diff
@@ -1007,19 +1149,486 @@ func (s) TestChainEngine(t *testing.T) {
+						},
+					},
+					wantStatusCode: codes.OK,
+					wantAuditEvents: []*audit.Event{
+						{
+							FullMethodName: "",
+							Principal:      "0.0.0.0",
+							PolicyName:     "test_policy",
+							MatchedRule:    "certain-source-ip",
+							Authorized:     true,
+						},
+					},
+				},
+				// This RPC should match with both the allow policy and deny policy
+				// and thus shouldn't be allowed to proceed as matched with deny.
+				// Audit logging matches with deny and short circuits
+				{
+					rpcData: &rpcData{
+						fullMethod: "localhost-fan-page",
+						peerInfo: &peer.Peer{
+							Addr: &addr{ipAddress: "0.0.0.0"},
+						},
+					},
+					wantStatusCode: codes.PermissionDenied,
+					wantAuditEvents: []*audit.Event{
+						{
+							FullMethodName: "localhost-fan-page",
+							Principal:      "0.0.0.0",
+							PolicyName:     "test_policy",
+							MatchedRule:    "localhost-fan",
+							Authorized:     false,
+						},
+					},
+				},
+				// This RPC shouldn't match with either policy, and thus
+				// shouldn't be allowed to proceed as didn't match with allow.
+				// Audit logging matches with the allow policy
+				{
+					rpcData: &rpcData{
+						peerInfo: &peer.Peer{
+							Addr: &addr{ipAddress: "10.0.0.0"},
+						},
+					},
+					wantStatusCode: codes.PermissionDenied,
+					wantAuditEvents: []*audit.Event{
+						{
+							FullMethodName: "",
+							Principal:      "10.0.0.0",
+							PolicyName:     "test_policy",
+							MatchedRule:    "",
+							Authorized:     false,
+						},
+					},
+				},
+				// This RPC shouldn't match with allow, match with deny, and
+				// thus shouldn't be allowed to proceed.
+				// Audit logging will have the deny logged
+				{
+					rpcData: &rpcData{
+						fullMethod: "localhost-fan-page",
+						peerInfo: &peer.Peer{
+							Addr: &addr{ipAddress: "10.0.0.0"},
+						},
+					},
+					wantStatusCode: codes.PermissionDenied,
+					wantAuditEvents: []*audit.Event{
+						{
+							FullMethodName: "localhost-fan-page",
+							Principal:      "10.0.0.0",
+							PolicyName:     "test_policy",
+							MatchedRule:    "localhost-fan",
+							Authorized:     false,
+						},
+					},
+				},
+			},
+		},
 	}
+	b := TestAuditLoggerBufferBuilder{}
+	audit.RegisterLoggerBuilder(b)
+	b2 := TestAuditLoggerCustomConfigBuilder{}
+	audit.RegisterLoggerBuilder(b2)
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
 			// Instantiate the chainedRBACEngine with different configurations that are
 			// interesting to test and to query.
-			cre, err := NewChainEngine(test.rbacConfigs)
+			cre, err := NewChainEngine(test.rbacConfigs, test.policyName)
 			if err != nil {
 				t.Fatalf("Error constructing RBAC Engine: %v", err)
 			}
 			// Query the created chain of RBAC Engines with different args to see
 			// if the chain of RBAC Engines configured as such works as intended.
 			for _, data := range test.rbacQueries {
 				func() {
+					auditEvents = nil
```

Comment:
Instead of doing this, I suggest you register and deregister the builder for each test case. You can thus keep a handle of the builder, where you can put this auditEvents. The builder can pass this along to the logger it creates.

Reply:
Done, PTAL because this was a more meaningful change